[HPB-22] [BE] post.status 評估使用 enum 管理

fufuYang opened the pull request:
## Description
### 測試案例編號：TC001
測試名稱： PostStatus Enum 正確定義與使用
測試對象： App\Enums\PostStatus
初始條件： 系統已存在 PostStatus Enum，包含 draft 與 published
測試步驟：
1. 檢查 Enum 中是否包含 Draft, Published 兩個值

預期結果：
1. Enum 正確定義，語法無誤
2. 每個值的 value 與名稱皆正確可用

✅ 測試結果：通過

### 測試案例編號：TC002
測試名稱： PostFactory 可正確產生 Enum status 資料
測試對象： PostFactory
初始條件： 已修改 Factory 使用 Enum
測試步驟：
1. 執行 Post::factory()->create()
2. 檢查產生出來的資料，其 status 欄位為 PostStatus::Published
3. 確保資料存入資料庫為 'published' 字串

預期結果：
1. Factory 可正常使用 Enum 產生資料
2. 資料轉換為正確的字串值儲存至資料庫

✅ 測試結果：通過

### 測試案例編號：TC003
測試名稱： 前端 Resource 表單正確顯示 Enum 選項
測試對象： Resource 表單中 status 欄位
初始條件： 表單已設定 ->options() 並使用 Enum
測試步驟：
1. 開啟 Resource 的建立文章頁面
2. 檢查 status 欄位選項是否為：Draft、Published
3. 檢查預設值為 draft

預期結果：
1. 選項來自 Enum 的 value
2. 表單顯示與預設值皆符合設計

✅ 測試結果：通過

### 測試案例編號：TC004
測試名稱： PostResource 顯示欄位正確使用 Enum
測試對象： Resource 頁面
初始條件： 顯示使用 Badge 或 Select 取自 Enum 
測試步驟：
1. 開啟文章列表頁
2. 檢查每筆資料的 status 欄位為 Draft 或 Published

預期結果：
1. 顯示欄位有 Draft 或 Published

✅ 測試結果：通過

## Related Task
[[HPB-22][BE] post.status 評估使用 enum 管理](https://www.notion.so/jyu1999/BE-post-status-enum-1fccba948f1180f08f39f6f51462d3ab)

fufuYang committed:
feat: 使用 PostStatus Enum 管理 post.status 欄位

- 新增 app/Enums/PostStatus.php 定義狀態 enum
- 調整 app/Models/Post.php，加上 cast: 'status' => PostStatus::class
- 調整 database/factories/PostFactory.php，使用 enum 產生狀態資料
- 調整 PostResource.php，Select 與 Badge 顯示改用 enum value 與 labels
- 修改 Controller，查詢與判斷邏輯改用 PostStatus Enum 處理

fufuYang committed:
refactor: 移除 post.status casts 與調整預設值為 PostStatus::Draft

- 移除 Post 模型中的 casts 配置，避免與 enum 實作衝突
- 修改 PostResource 中的預設值為 PostStatus::Draft 以對應 enum 使用方式"

JYu1999 commented on app/Filament/Resources/PostResource.php:
```diff
@@ -44,11 +45,11 @@ public static function form(Form $form): Form
-                        'draft' => 'Draft',
-                        'published' => 'Published',
+                        PostStatus::Draft->value => 'Draft',
+                        PostStatus::Published->value => 'Published',
                     ])
                     ->required()
-                    ->default('draft')
+                    ->default(PostStatus::Draft)
```
有加上 `->value` 跟沒有差在哪裡？

fufuYang replied:
PostStatus::Draft 本身是一個 enum 物件，->value 的作用是取得 enum 實際對應的值。

當初會這樣寫，是因為在 Post model 中有設定 casts: 'status' => PostStatus::class，因此可以直接使用 enum 實例（ PostStatus::Draft）作為預設值或查詢條件，Laravel 會協助自動轉型。

不過後來因為 casts 設定會與顏色 BadgeColumn 的顯示邏輯衝突，最後決定移除 casts 設定。

所以現在在表單中，->default() 的寫法也應調整為 ->default(PostStatus::Draft->value)，以確保預設值為字串 'draft'，而不是 enum object，避免資料庫儲存錯誤型別。

fufuYang replied:
已經完成修改，統一使用了 PostStatus::Draft->value

fufuYang commented on app/Enums/PostStatus.php:
```diff
@@ -0,0 +1,9 @@
+
+namespace App\Enums;
+
+enum PostStatus: string
+{
+    case Draft = 'draft';
+    case Published = 'published';
+}
```
我思考的部份是
1. 使用 'draft'、'published' 能直觀的表達狀態，可讀性高
2. 資料表型態為 VARCHAR ，能直接使用文字儲存 draft / published，就不需要查表對照。

JYu1999 replied:
1. 變數名稱明確就好，有必要連變數內容都要很直觀嗎？
2. 那有考慮過資料庫也改存 int 嗎？（不能用現在有的東西當成佐證，現有的東西不見得是合理的）

fufuYang replied:
有考慮過用 int，但目前狀態數量少，對效能或儲存影響不大，所以覺得可以使用 string 讓資料更直觀易懂。
後續如果狀態變多或有效能需求，是可以轉換成 int。

JYu1999 replied:
ok

fufuYang committed:
fix: 修正 Select 預設值為字串，避免 enum instance 自動轉型問題

- 將 ->default(PostStatus::Draft) 改為 ->default(PostStatus::Draft->value)

fufuYang committed:
feat: 修正 PostStatus 使用方式

將 PostStatus::Published 更新為 PostStatus::Published->value，確保在 Post Factory 和 Post Controller 中都能正確處理 PostStatus Enum

fufuYang committed:
Merge branch 'main' into feature/HPB-22

JYu1999 commented on app/Filament/Resources/PostResource.php:
```diff
@@ -44,11 +45,11 @@ public static function form(Form $form): Form
                     ->options([
-                        'draft' => 'Draft',
-                        'published' => 'Published',
+                        PostStatus::Draft->value => 'Draft',
+                        PostStatus::Published->value => 'Published',
```
我其實在想什麼時候要用 `enum->value`，什麼時候用 `enum->name`。

比如說我看到這裡就在想，為什麼不是：

```php
<?php
...
PostStatus::Draft->name => PostStatus::Draft->value,
...
```

fufuYang replied:
我目前的理解是
`enum->name` 通常用來作為下拉選單上「顯示給使用者看的文字」
`enum->value` 則是實際儲存進資料庫的值。
所以會根據使用的場景，決定要取哪個屬性。

看到結語提供的想法，我在想或許可以在 `enum` 裡新增一個靜態方法 `toSelectArray()`，這樣在 `Select::make()` 裡，就可以直接用 `PostStatus::toSelectArray()` 來取得狀態陣列。

好處是之後如果有新增或修改狀態，只要改 `enum` 本身就好，不需要到每個 `options()` 的地方都手動維護。
可以讓狀態選項的維護更集中，也減少重複程式碼。

```php
public static function toSelectArray(): array
{
    $options = [];
    foreach (self::cases() as $case) {
        $options[$case->value] = $case->name;
    }
    return $options;
}
```

JYu1999 replied:
我覺得你要想一個狀況是，有時候我們想讓使用者看到的不是 `enum->name`。

假設我想要讓狀態為 `draft` 的文章，使用者看到 `草稿`， 用你目前的做法就行不通。

而從根本上，`enum->name` 代表的是變數名稱，我們本來就不會希望使用者看到變數名稱

fufuYang replied:
好，了解了，這我會再去思考。
依目前的情境我覺得現在的寫法就可以了。

fufuYang committed:
feat(posts): 建立 migration(2025_07_27_063526_alter_posts_status_column.php)，將 posts 表的 status 欄位預設值設定為 PostStatus::Draft。

fufuYang commented:
@JYu1999 
我已經將 Post.status 相關的部分都已替換成 Enum，並進行以下測試

### 測試案例編號：TC005
測試名稱： 確認 posts 表 status 欄位預設值為 PostStatus::Draft
測試對象： posts 資料表的 status 欄位
初始條件： 已透過遷移將 status 欄位預設值設為 PostStatus::Draft。
測試步驟：

1. 本地網頁建立文章： 在建立文章頁面不指定 status 欄位並送出。
2. 工廠模式建立文章： 執行 Post::factory()->create()，不指定 status 欄位。

預期結果：

1. 無論透過網頁或工廠模式，新增文章時，其 status 欄位實際存入的值應為 draft 字串。

✅ 測試結果：通過
